chia/full_node/full_node_api.py

```diff
@@ -1426,50 +1425,28 @@ async def request_puzzle_solution(self, request: wallet_protocol.RequestPuzzleSo
-        """Returns header blocks by directly streaming bytes into Message
-
+        """
         This method should be used instead of RequestHeaderBlocks
         """
         reject = RejectBlockHeaders(request.start_height, request.end_height)
-
         if request.end_height < request.start_height or request.end_height - request.start_height > 128:
             return make_msg(ProtocolMessageTypes.reject_block_headers, reject)
-        if self.full_node.block_store.db_wrapper.db_version == 2:
-            try:
-                blocks_bytes = await self.full_node.block_store.get_block_bytes_in_range(
-                    request.start_height, request.end_height
-                )
-            except ValueError:
-                return make_msg(ProtocolMessageTypes.reject_block_headers, reject)
-
-        else:
-            height_to_hash = self.full_node.blockchain.height_to_hash
-            header_hashes: list[bytes32] = []
-            for i in range(request.start_height, request.end_height + 1):
-                header_hash: Optional[bytes32] = height_to_hash(uint32(i))
-                if header_hash is None:
-                    return make_msg(ProtocolMessageTypes.reject_header_blocks, reject)
-                header_hashes.append(header_hash)
-
-            blocks_bytes = await self.full_node.block_store.get_block_bytes_by_hash(header_hashes)
-        if len(blocks_bytes) != (request.end_height - request.start_height + 1):  # +1 because interval is inclusive
+        try:
+            header_blocks_map = await self.full_node.blockchain.get_header_blocks_in_range(
+                request.start_height, request.end_height, request.return_filter
+            )
+        except ValueError:
+            return make_msg(ProtocolMessageTypes.reject_block_headers, reject)
+        if len(header_blocks_map) != (
+            request.end_height - request.start_height + 1
+        ):  # +1 because interval is inclusive
             return make_msg(ProtocolMessageTypes.reject_block_headers, reject)
-        return_filter = request.return_filter
-        header_blocks_bytes: list[bytes] = [header_block_from_block(memoryview(b), return_filter) for b in blocks_bytes]
-
-        # we're building the RespondHeaderBlocks manually to avoid cost of
-        # dynamic serialization
-        # ---
-        # we start building RespondBlockHeaders response (start_height, end_height)
-        # and then need to define size of list object
-        respond_header_blocks_manually_streamed: bytes = (
-            uint32(request.start_height).stream_to_bytes()
-            + uint32(request.end_height).stream_to_bytes()
-            + uint32(len(header_blocks_bytes)).stream_to_bytes()
+        return make_msg(
+            ProtocolMessageTypes.respond_block_headers,
+            wallet_protocol.RespondBlockHeaders(
+                request.start_height, request.end_height, list(header_blocks_map.values())
+            ),
         )
```

Comment:
if you're removing this optimization, you need to demonstrate (with a benchmark) that the performance degradation is acceptable. The reason we did this was because a node could be quite bogged down by responding to these requests

Reply:
I benchmarked the following 3 versions:

1. Original version, with the bug of not accounting for additions and removals:
```
with benchmark_runner.assert_runtime(
    print=False, seconds=1, label="header_block_from_block without additions and removals"
):
    [block_bytes] = await full_node_api.full_node.block_store.get_block_bytes_by_hash([new_block.header_hash])
    header_block_bytes = header_block_from_block(memoryview(block_bytes))
```

2. Original version, fixed to account for additions and removals:
```
with benchmark_runner.assert_runtime(
    print=False, seconds=1, label="header_block_from_block with additions and removals"
):
    [block_bytes] = await full_node_api.full_node.block_store.get_block_bytes_by_hash([new_block.header_hash])
    added_coins_records, removed_coins_records = await asyncio.gather(
        full_node_api.full_node.coin_store.get_coins_added_at_height(uint32(new_block.height)),
        full_node_api.full_node.coin_store.get_coins_removed_at_height(uint32(new_block.height)),
    )
    tx_addition_coins = [cr.coin for cr in added_coins_records if not cr.coinbase]
    removal_names = [cr.coin.name() for cr in removed_coins_records]
    header_block_bytes = header_block_from_block(
        memoryview(block_bytes), tx_addition_coins=tx_addition_coins, removal_names=removal_names
    )
```
3. The (simpler) version of this PR:
```
with benchmark_runner.assert_runtime(print=False, seconds=1, label="get_header_blocks_in_range version"):
    header_blocks_map = await full_node_api.full_node.blockchain.get_header_blocks_in_range(
        uint32(new_block.height), uint32(new_block.height), True
    )
    result = list(header_blocks_map.values())
```

Benchmarking across 50 runs gives us the following average durations:
```
1. Avg run for header_block_from_block without additions and removals: 0.0005007242900000142
2. Avg run for header_block_from_block with additions and removals: 0.0005900122299999877
3. Avg run for get_header_blocks_in_range version: 0.0003575485899999675
```